protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
+
+Note that the above functionality also makes for a good set of milestones, each of which can and
+should be implemented incrementally.
+
+## Declarative storage management
+
+Taking inspiration from Chugsplash, we wish to separate the concern of what bytecode to run, and
+what to write to storage. However we also wish to avoid using the current `StorageSetter` pattern,
+which simply provides logic to "write anything anywhere" in storage.
+
+In order to achieve this, each contract will be refactored to inherit from a `__Layout` contract. By
+way of example, we take an initializable version of `SimpleStorage`, and refactor it into the set of
+contracts which would be required to specify a hypothetical Magma release, which adds a new variable
+to the contract.
+
+```solidity
+contract SimpleStorage is Initializable {
+  uint public foo;
+  bytes32 public bar; // new variable added for the Magma release.
+
+
+  /// @notice initialize is called whether for a fresh system or when upgrading an existing system.
+  function initialize(uint _foo, bytes32 bar) public initializer {
+    // chains being upgraded need to read the current value of foo, then write it again.
+    foo = _foo;
+    bar = _bar;
+  }
+}
+```
+
+This `SimpleStorage` will become:
+
+```solidity
+abstract contract SimpleStorageLayout {
+  uint public foo;
+  bytes32 public bar;
+}
+
+contract SimpleStoragePopulator is SimpleStorageLayout {
+  function initialize() public {
+    foo = 1;
+    bar = 2_000
+  }
+
+  function upgrade() public {
+    bar = 2_000;
+  }
+}
+
+contract SimpleStorage is SimpleStorageLayout {
+  // Whatever runtime logic is needed
+}
+```
+
+With this architecture, the OPCM contract (which will be a non-proxied singleton per release) would
+then either:
+
+1. Deploy a new contract by:
+   1. Deploying a new proxy
+   1. setting its implementation to the `SimpleStoragePopulator`
+   1. calling `initialize()`
+   1. setting its implementation to `SimpleStorage`
+1. Upgrade a contract by:
+   1. setting its implementation to the `SimpleStoragePopulator`
+   1. calling `upgrade()`
+   1. setting its implementation to `SimpleStorage`
+
+This workflow has the benefit of removing the `initializer()` functions, which:
+
+- create a risk of reinit attacks
+- bloat the runtime bytecode size
+- are extremely verbose
+- require finesse to avoid modifying or resetting values which are not meant to be changed.
+
+## A single `SuperchainProxyAdmin`
+
+The existing L1 `ProxyAdmin` (of which there is currently one per OP Chain), will be replaced with a
+single L1 `SuperchainProxyAdmin`. The `SuperchainProxyAdmin`, in contrast with the L1 `ProxyAdmin`,
```

Comment:
Just to avoid any confusion:
- Technically there are 2 roles that have a proxy admin: The SuperchainConfig and ProtocolVersions proxies, and the `Ownable` parts of contracts like the SystemConfig and DisputeGameFactory (DGF)
- Pre-OPCM, for the standard config, we use the same exact proxy admin for both of those, conflating the two roles
- OPCM split these out into two distinct proxy admins: a `superchainProxyAdmin` and an `opChainProxyAdmin`

Therefore I just want to avoid confusion on what `SuperchainProxyAdmin` means here since it's basically talking about a new version of the `ProxyAdmin` contract

Reply:
Agree.